File: RUST/sfo_06_print_a_linked_list_reverse_order_s1/src/main.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

pub struct Solution {}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn reverse_print(head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut ans = Vec::new();
        let mut res = Vec::new();
        let mut p = &head;
        while let Some(node) = p {
            ans.push(node.val);
            p = &node.next;
        }
        while let Some(v) = ans.pop() {
            res.push(v);
        }
        res
    }
}

fn vec_to_linked_list(nums: Vec<i32>) -> Option<Box<ListNode>> {
    let mut dummy = ListNode::new(0);
    let mut current = &mut dummy;

    for &num in nums.iter() {
        current.next = Some(Box::new(ListNode::new(num)));
        current = current.next.as_mut().unwrap();
    }

    dummy.next
}

fn main() {
    let input = vec![1, 3, 2];
    println!("{:?}", input);
    let head = vec_to_linked_list(input);
    let ans = Solution::reverse_print(head);

    println!("{:?}", ans);
}

File: CPP/sfo_06_print_a_linked_list_reverse_order_s1/main.cc
#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    vector<int> reversePrint( ListNode* head ) {
        while ( head != NULL ) {
            s1.push( head->val );
            head = head->next;
        }
        vector<int> res;
        while ( !s1.empty() ) {
            res.push_back( s1.top() );
            s1.pop();
        }
        return res;
    }

private:
    stack<int> s1;
};

int main() {
    // ======= Test Case =======
    ListNode* head = vectorToLinkedList( vector<int>{ 1, 3, 2 } );
    // ====== Driver Code ======
    Solution*   slt = new Solution();
    vector<int> res = slt->reversePrint( head );
    PrintUtil::printVector( res );

    return 0;
}
